### PR TITLE
Bug 799734 - BrowserID crypto library

### PR DESCRIPTION
**Pull request by ncalexan**
Pretty much just enough for MockMyID.com.  Only does RSA crypto.  But it needs to land in some form soon.

### CODE REVIEW

**Comment by ncalexan** [comment]
After you review https://github.com/mozilla-services/android-sync/pull/270 this can be rebased down to 3 really rather small commits!

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
+package org.mozilla.gecko.browserid.crypto;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:1]: MPL.

**Comment by rnewman** [comment]
Your call whether this should be o.m.g.browserid or something else (Persona?).

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  public String getAlgorithm() {
+    return jwTokenFactory.getAlgorithm();
+  }
+
+  public JWTokenFactory getJWTokenFactory() {
+    return jwTokenFactory;
+  }
+
+  protected String getPayloadString(
+      String payloadString,
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:32]: Aha, I cunningly already provide the payload string. Err…

I think you mean "`jsonContents`" or something.

Actually, you probably want an overloaded version of this that takes an `ExtendedJSONObject`, no?

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+      payload.put("aud", audience);
+    }
+    payload.put("exp", expiresAt);
+
+    return payload.toJSONString();
+  }
+
+  // protected abstract String getHeaderString();
+
+//  public static class JWSignature {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:57]: Leave it in or take it out.

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+//      this.signatureBytes = Base64.decodeBase64(signature);
+//    }
+//  }
+
+  public String sign(
+      String payloadString,
+      String issuer,
+      long issuedAt,
+      String audience,
+      long expiresAt,
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:101]: Cluster fields by type.

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+//
+//    String string = Base64.encodeBase64URLSafeString(header.getBytes("UTF-8")) +
+//        "." +
+//        Base64.encodeBase64URLSafeString(payload.getBytes("UTF-8"));
+//
+//    byte[] signatureBytes = jwTokenFactory.encode(payload, privateKey, null);
+//
+//    return new JWSignature(header, payload, signatureBytes);
+
+    return jwTokenFactory.encode(payload, privateKey);
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:115]: You haven't written a comment saying that these methods must be called from only one thread at a time, which means you're permitting concurrent access to `jwTokenFactory`.

If the token factory is thread-safe, make it a singleton and have these be static methods.

If it's not, you need to `synchronize` these methods, at which point you should consider stuffing more state into the object if it makes sense.

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+  public final String getCertificatePayloadString(
+      PublicKey publicKeyToSign,
+      String email) throws Exception {
+    ExtendedJSONObject principal = new ExtendedJSONObject();
+    principal.put("email", email);
+
+    ExtendedJSONObject payload = new ExtendedJSONObject();
+    payload.put("principal", principal);
+    payload.put("public-key", new ExtendedJSONObject(publicKeySerializer.serializePublicKey(publicKeyToSign)));
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:126]: Break this line out into well-named locals.

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    String payloadString = getPayloadString(certificatePayloadString, issuer, issuedAt, null, expiresAt);
+
+    return jwTokenFactory.encode(payloadString, privateKey);
+  }
+
+  public String createAssertion(
+      PrivateKey privateKeyToSignWith,
+      String certificate,
+      String issuer,
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:148]: It seems like there's a bunch of state that's shared between these calls — issuer, audience, a key pair, etc.

Does it make sense to you to bundle these into a constructed object that exposes assertion/cert/etc. generation?

File: src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    return certificate + "~" + signature;
+  }
+
+  public String createAssertion(
+      PrivateKey privateKeyToSignWith,
+      String certificate,
+      String issuer,
+      String audience) throws Exception {
+    long issuedAt = System.currentTimeMillis();
+    long durationInMilliseconds = 60 * 60 * 1000;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/crypto/BrowserIDTokenFactory.java:179]: Lift default duration into a constant.

File: src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.spec.InvalidKeySpecException;
+
+import org.mozilla.gecko.browserid.crypto.BrowserIDTokenFactory;
+import org.mozilla.gecko.jwcrypto.RSACryptoImplementation;
+
+public class MockMyIDTokenFactory {
+  public static final String     MOCKMYID_ALGORITHM        = "RS256";
+  public static final BigInteger MOCKMYID_MODULUS          = new BigInteger("15498874758090276039465094105837231567265546373975960480941122651107772824121527483107402353899846252489837024870191707394743196399582959425513904762996756672089693541009892030848825079649783086005554442490232900875792851786203948088457942416978976455297428077460890650409549242124655536986141363719589882160081480785048965686285142002320767066674879737238012064156675899512503143225481933864507793118457805792064445502834162315532113963746801770187685650408560424682654937744713813773896962263709692724630650952159596951348264005004375017610441835956073275708740239518011400991972811669493356682993446554779893834303");
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java:19]: Comment where these came from, when they might change, etc.

File: src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+   * @throws Exception
+   */
+  public String createMockMyIDCertificate(final PublicKey publicKeyToSign, final String username,
+      final long issuedAt, final long durationInMilliseconds)
+      throws Exception {
+    final long expiresAt = issuedAt + durationInMilliseconds;
+
+    PrivateKey mockMyIdPrivateKey = getMockMyIDPrivateKey();
+
+    return browserIdTokenFactory.createCertificate(publicKeyToSign, username + "@mockmyid.com", "mockmyid.com", issuedAt, expiresAt, mockMyIdPrivateKey);
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java:64]: Named variables for these things, please:

```
String authority = "mockmyid.com";
String whatever = username + "...";
return factory.cre…(…, authority, …);
```

Also, constants!

File: src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+   *          public key to sign.
+   * @param username
+   *          sign username@mockmyid.com
+   * @return Java Web Signature.
+   * @throws Exception
+   */
+  public String createMockMyIDCertificate(final PublicKey publicKeyToSign, final String username)
+      throws Exception {
+    return createMockMyIDCertificate(publicKeyToSign, username,
+        System.currentTimeMillis(), 60 * 60 * 1000);
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java:81]: This is a magic duration. Should it be shared with the other use of this magic duration?

File: src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    final long expiresAt = issuedAt + durationInMilliseconds;
+
+    PrivateKey mockMyIdPrivateKey = getMockMyIDPrivateKey();
+
+    return browserIdTokenFactory.createCertificate(publicKeyToSign, username + "@mockmyid.com", "mockmyid.com", issuedAt, expiresAt, mockMyIdPrivateKey);
+  }
+
+  /**
+   * Sign a public key asserting ownership of username@mockmyid.com with
+   * mockmyid.com's private key.
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/mockmyid/MockMyIDTokenFactory.java:69]: Mention duration.

File: src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDRemoteVerifierClient.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    BaseResource r = new BaseResource(verifierUri);
+
+    r.delegate = new BaseResourceDelegate(r) {
+      @Override
+      public void handleHttpResponse(HttpResponse response) {
+        SyncResponse res = new SyncResponse(response);
+        int statusCode = res.getStatusCode();
+        Logger.debug(LOG_TAG, "Got response with status code " + statusCode + ".");
+
+        ExtendedJSONObject o = null;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDRemoteVerifierClient.java:65]: `final` and don't assign `null`. You always return in the `catch` block, so the compiler can guarantee assignment.

File: src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDRemoteVerifierClient.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+          delegate.handleError(new BrowserIDVerifierMalformedResponseException(e));
+          return;
+        }
+
+        if (statusCode != 200) {
+          delegate.handleError(new BrowserIDVerifierErrorResponseException("Expected status code 200."));
+          return;
+        }
+
+        if ("failure".equals(o.getString("status"))) {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDRemoteVerifierClient.java:78]: `STATUS_FAILURE`.

File: src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDRemoteVerifierClient.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+          delegate.handleFailure(o.getString("reason"));
+          return;
+        }
+
+        if (!("okay".equals(o.getString("status")))) {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDRemoteVerifierClient.java:83]: `STATUS_OK`.

File: src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDVerifierException.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
+package org.mozilla.gecko.browserid.verifier;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/browserid/verifier/BrowserIDVerifierException.java:1]: MPL.

File: src/main/java/org/mozilla/gecko/jwcrypto/JWCryptoImplementation.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
+package org.mozilla.gecko.jwcrypto;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/jwcrypto/JWCryptoImplementation.java:1]: MPL.

File: src/main/java/org/mozilla/gecko/jwcrypto/JWTokenFactory.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
+package org.mozilla.gecko.jwcrypto;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/jwcrypto/JWTokenFactory.java:1]: MPL.

File: src/main/java/org/mozilla/gecko/jwcrypto/JWTokenFactory.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+  public String encode(String payload, PrivateKey privateKey) throws Exception {
+    return encode(payload, privateKey, null);
+  }
+
+  protected String encode(String payload, PrivateKey privateKey, Map<String, Object> headerFields) throws Exception {
+    ExtendedJSONObject header = new ExtendedJSONObject();
+    if (headerFields != null) {
+      for (Entry<String, Object> entry : headerFields.entrySet()) {
+        header.put(entry.getKey(), entry.getValue());
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/jwcrypto/JWTokenFactory.java:37]: Maybe add `putAll` to `ExtendedJSONObject`, if it isn't there already?

File: src/main/java/org/mozilla/gecko/jwcrypto/PublicKeySerializer.java [hunk 20]
```diff
@@ -0,0 +1,18 @@
+package org.mozilla.gecko.jwcrypto;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/jwcrypto/PublicKeySerializer.java:1]: MPL.

File: src/main/java/org/mozilla/gecko/jwcrypto/RSACryptoImplementation.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
+package org.mozilla.gecko.jwcrypto;
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/jwcrypto/RSACryptoImplementation.java:1]: MPL.

File: src/test/java/org/mozilla/gecko/browserid/crypto/test/TestBrowserIDTokenFactory.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
+package org.mozilla.gecko.browserid.crypto.test;
```
Comment (rnewman) [inline src/test/java/org/mozilla/gecko/browserid/crypto/test/TestBrowserIDTokenFactory.java:1]: MPL.

File: src/test/java/org/mozilla/gecko/jwcrypto/test/TestJWTokenFactory.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
+package org.mozilla.gecko.jwcrypto.test;
```
Comment (rnewman) [inline src/test/java/org/mozilla/gecko/jwcrypto/test/TestJWTokenFactory.java:1]: License.

**Comment by rnewman** [comment]
Some minor tidying and refactoring that needs to take place, but it's in a good direction, I think.

You should get crypto code review from the sec team, because I'm not qualified to do that, or indeed to verify that this correctly implements BrowserID's crypto. Flag the bug, ask `curtisko` to assign someone to help (`dchan`?).

**Comment by ncalexan** [comment]
This is looking less and less likely to be used, so I'm going to close this pull request with a comment in the bug linking to the review comments.

**Pull request closed without merging**